### DISCUSSION TITLE
Stability of the `enum-map` impl

### DISCUSSION THREAD

**Discussion by aumetra** [greyblake/whatlang-rs#133] [Q&A] [answered]
I would like to use `whatlang` for language detection and store its detected language as an integer into a database (to just save that little bit of storage).

What are the stability guarantees around the `enum-map` implementation?  
Will the value returned by `into_usize` always stay the same for a particular language or is it prone to change?

**Comment by greyblake** [comment, marked as answer]
Hi,

you don't need `enum-map` for that. It would be misuse.

You can convert `whatlang::Lang` to an integer by simple casting:

```rs
use whatlang::Lang;

fn main() {
    let ukrainian_index = Lang::Ukr as i32;
    println!("ukranian_index = {ukrainian_index}");

    let esperanto_index = Lang::Epo as i32;
    println!("esperanto_index = {esperanto_index}");
}
```
Output:
```
ukranian_index = 10
esperanto_index = 0
```

This works, but I cannot give you guarantee, that this won't change in the next minor version (e.g. 0.17.0). The indices my rearrange a bit if we add a new language.

**So I would strongly recommend NOT rely on this behavior if you want to persist languages in database**.

If you need to persist language in DB I'd recommend using ISO 639-3 codes. In DB you can use enums (e.g. if you're using Postgres). or just VARCHAR, string.

Whatlang provides functions to convert language to code and back:

```rs
use whatlang::Lang;

fn main() {
    let code = Lang::Ukr.code();
    println!("code = {code}");

    let lang = Lang::from_code("epo").unwrap();
    println!("lang = {}", lang.name());
}
```

Output:

```
code = ukr
lang = Esperanto
```

**Reply by aumetra** [reply]
Ah I was just thinking about saving a few bytes per entry. I'll take the ISO code route then, thanks!

**Reply by greyblake** [reply]
I don't know which DB you're using, but if you're using PostgreSQL, you can use PostgreSQL's enums.
It will occupy the same space as an integer.

Using pure integers, would be also not very friendly for potential BI, data analysis.

**Reply by aumetra** [reply]
Yeah that's what I'm thinking about now. Autogenerating the SQL enum declaration and applying it on start-up (I'm on Postgres)
